src/air/applications.py

```diff
@@ -380,6 +391,329 @@ def about_us(): # routes is "/about-us"
+            Doc(
+                """
+                Customize the function used to generate unique IDs for the *path
+                operations* shown in the generated OpenAPI.
+
+                This is particularly useful when automatically generating clients or
+                SDKs for your API.
+
+                Read more about it in the
+                [FastAPI docs about how to Generate Clients](https://fastapi.tiangolo.com/advanced/generate-clients/#custom-generate-unique-id-function).
+                """
+            ),
+        ] = default_generate_unique_id,
+    ) -> None:
+        """
+        Include an `APIRouter` in the same app.
+
+        Read more about it in the
+        [FastAPI docs for Bigger Applications](https://fastapi.tiangolo.com/tutorial/bigger-applications/).
+
+        ## Example
+
+        ```python
+        from fastapi import FastAPI
+
+        from .users import users_router
+
+        app = FastAPI()
+
+        app.include_router(users_router)
+        ```
+        """
+        super().include_router(
+            router,
+            prefix=prefix,
+            tags=tags,
+            dependencies=dependencies,
+            responses=responses,
+            deprecated=deprecated,
+            include_in_schema=include_in_schema,
+            default_response_class=default_response_class,
+            callbacks=callbacks,
+            generate_unique_id_function=generate_unique_id_function,
+        )
+
+        # when including routers, also (re)bind .url so prefixes are respected
+        self._attach_url_descriptors()
+
+    def _attach_url_descriptors(self) -> None:
+        """
+        Bind reverse-URL helpers to all registered HTTP endpoints.
+
+        Walks `self.router.routes` and, for each `APIRoute` that has a `name` and a
+        callable `endpoint`, ensures the endpoint has a `.url(**params) -> str`
+        attribute (via `UrlDescriptor`).
+        """
+        for route in self.router.routes:
+            if isinstance(route, routing.APIRoute) and route.name and callable(route.endpoint):
+                self._attach_url_descriptor_to_route(route.name, route.endpoint)
+
+    def _attach_url_descriptor_to_route(self, route_name: str, endpoint: Callable[..., Any]) -> None:
+        """
+        Attach a `.url` callable to a single endpoint, if possible.
+
+        Adds `endpoint.url = UrlDescriptor(self, route_name)` so handlers can build
+        links like `handler.url(id=123, ref="home")`. If `.url` already exists, returns immediately.
+        """
+        attr = "url"
+        if hasattr(endpoint, attr):
+            return  # already attached
+        if inspect.isfunction(endpoint) or inspect.ismethod(endpoint) or hasattr(endpoint, "__dict__"):
+            setattr(endpoint, attr, UrlDescriptor(self, route_name))
+        else:
+            # We can't attach an attribute (e.g. callable instance with __slots__)
+            # better to have some warning logs here ...
```

Comment:
Oof - this is the logic branch we should track early on to see what's happening.